firedrake/assign.py

```diff
@@ -0,0 +1,130 @@
+import abc
+
+import numpy as np
+from pyop2.utils import cached_property
+import ufl
+from ufl.algorithms import extract_coefficients
+from ufl.constantvalue import RealValue, Zero
+from ufl.corealg.map_dag import map_expr_dag
+from ufl.corealg.multifunction import MultiFunction
+
+from firedrake.utils import ScalarType
+from firedrake.vector import Vector
+
+
+class WeightedCoefficientCollector(MultiFunction):
+    def __init__(self):
+        super().__init__()
+        self.weighted_coefficients = []
+
+    def sum(self, o):
+        for op in o.ufl_operands:
+            self(op)
+
+    def product(self, o):
+        try:
+            weight, = filter(
+                lambda op: isinstance(op, (RealValue, Zero)), o.ufl_operands
+            )
+            coeff, = filter(lambda op: isinstance(op, ufl.Coefficient), o.ufl_operands)
+        except ValueError:
+            raise ValueError("Only products of scalars and coefficients allowed")
+        self.weighted_coefficients.append((weight, coeff))
+
+    def coefficient(self, o):
+        self.weighted_coefficients.append((1, o))
+
+    def invalid(self, o):
+        raise ValueError(f"{o} is not a valid component of a weighted sum of coefficients")
```

Comment:
Does this mean that `f.assign(as_tensor([[1, 2], [3, 4]]) @ as_vector([a, b]))` will be rejected? That is a linear combination so should be safe.

Reply:
Running this gives me an `unsupported operand type(s) for @` error. Are you sure that UFL supports this sort of thing?

Reply:
Ah ok, I see.